app/models/task.rb

```diff
@@ -19,10 +19,27 @@ class Task < ApplicationRecord
   belongs_to :supplier
 
   has_many :submissions, dependent: :nullify
-  has_one :latest_submission, -> { order(created_at: :desc) }, inverse_of: :task, class_name: 'Submission'
 
   delegate :name, to: :supplier, prefix: true
 
+  completed_or_latest_scope = lambda do
+    where("
+        CASE WHEN
+          EXISTS(
+            SELECT id
+              FROM submissions
+              WHERE aasm_state = 'completed' AND task_id = $1
+          )
+        THEN
+          aasm_state = 'completed'
+        ELSE
+          1=1
+        END
+    ").order(created_at: :desc)
```

Comment:
I *think* this can be done with a self-join rather than a `CASE` and `EXISTS` – just checking out branch to see

Reply:
Nope. Can't do (or rather can't do last thing on Thu :wink:)

Reply:
This is really quick and nice against a production sample